include/swift/ABI/Task.h

```diff
@@ -190,12 +210,362 @@ class AsyncTask : public HeapObject, public Job {
+
+        static const unsigned long maskPending    = 0x0000000FFFFF00000l;
+        static const unsigned long onePendingTask = 0x00000000000100000l;
+
+        static const unsigned long maskWaiting    = 0x000000000000FFFFFl;
+        static const unsigned long oneWaitingTask = 0x00000000000000001l;
+
+        unsigned long status;
+
+        unsigned int readyTasks() {
+          return (status & maskReady) >> 40;
+        }
+
+        unsigned int pendingTasks() {
+          return (status & maskPending) >> 20;
+        }
+
+        unsigned int waitingTasks() {
+          return status & maskWaiting;
+        }
+
+        bool isEmpty() {
+          return pendingTasks() == 0;
+        }
+
+        /// Status value decrementing the Ready, Pending and Waiting counters by one.
+        GroupStatus completingReadyPendingWaitingTask() {
+          assert(pendingTasks() > 0 && "can only complete waiting tasks when pending tasks available");
+          assert(readyTasks() > 0 && "can only complete waiting tasks when ready tasks available");
+          assert(waitingTasks() > 0 && "can only complete waiting tasks when waiting tasks available");
+          return GroupStatus { status - oneReadyTask - oneWaitingTask - onePendingTask };
+        }
+
+        /// Pretty prints the status, as follows:
+        /// GroupStatus{ P:{pending tasks} W:{waiting tasks} {binary repr} }
+        std::string to_string() {
+          std::string str;
+          str.append("GroupStatus{ ");
+          str.append("R:");
+          str.append(std::to_string(readyTasks()));
+          str.append(" P:");
+          str.append(std::to_string(pendingTasks()));
+          str.append(" W:");
+          str.append(std::to_string(waitingTasks()));
+          str.append(" " + std::bitset<64>(status).to_string());
+          str.append(" }");
+          return str;
+        }
+
+        /// Initially there are no waiting and no pending tasks.
+        static const GroupStatus initial() {
+          return GroupStatus { 0 };
+        };
+    };
+
+    template<typename T>
+    class NaiveQueue {
+        std::queue<T> queue;
+
+    public:
+        NaiveQueue() = default;
+        NaiveQueue(const NaiveQueue<T> &) = delete ;
+        NaiveQueue& operator=(const NaiveQueue<T> &) = delete ;
+
+        NaiveQueue(NaiveQueue<T>&& other) {
+          queue = std::move(other.queue);
+        }
+
+        virtual ~NaiveQueue() { }
+
+        bool dequeue(T &output) {
+          if (queue.empty()) {
+            return false;
+          }
+          output = queue.front();
+          queue.pop();
+          return true;
+        }
+
+        void enqueue(const T item) {
+          queue.push(item);
+        }
+
+    };
+
+  private:
+
+    // TODO: move to lockless via the status atomic
+    mutable std::mutex mutex;
+
+    /// Used for queue management, counting number of waiting and ready tasks
+    std::atomic<unsigned long> status;
```

Comment:
@DougGregor so all of this dance around the status and the two queues is remnants of attempts of the lockless impl; 

I'm planning to get back to it and making it happen now that John's runtime landed -- I did not know if you'd rather we remove all of it and add proper thing later; or simply keep status quo + the lock because it works just fine enough; Honestly I'd prefer keeping as is as then I can spare a cycle of restoring similar infra...

Let me know if you'd rather only merge a perfectly clean locking impl for now. I did remove all the other not currently used MPSC queues and stuff so at least we do not pollute with not used datastructures.